foqus_lib/framework/surrogate/scaling.py

```diff
@@ -0,0 +1,222 @@
+
+
+def scale_linear(array_in, lo=None, hi=None):
+    if lo is None:
+        lo = np.min(array_in)
+    if hi is None:
+        hi = np.max(array_in)
+    validate_for_scaling(array_in, lo, hi)
+    if (hi - lo) == 0:
+        result = 0
+    else:
+        result = (array_in - lo) / (hi - lo)
+    return result
+
+
+def scale_log(array_in, lo=None, hi=None):
+    # need to account for log domain
+    if np.any(array_in <= 0):
+        raise ValueError("All values must be > 0 to use scale_log")
+    if lo is None:
+        lo = np.min(array_in)
+    if hi is None:
+        hi = np.max(array_in)
+    validate_for_scaling(array_in, lo, hi)
+    result = (np.log10(array_in) - np.log10(lo)) / (np.log10(hi) - np.log10(lo))
+    return result
+
+
+def scale_log2(array_in, lo=None, hi=None):
+    if lo is None:
+        lo = np.min(array_in)
+    if hi is None:
+        hi = np.max(array_in)
+    validate_for_scaling(array_in, lo, hi)
+    result = np.log10(9 * (array_in - lo) / (hi - lo) + 1)
+    return result
+
+
+# fix expected values in test
+
+
+def scale_power(array_in, lo=None, hi=None):
+    if lo is None:
+        lo = np.min(array_in)
+    if hi is None:
+        hi = np.max(array_in)
+    validate_for_scaling(array_in, lo, hi)
+    result = (np.power(10, array_in) - np.power(10, lo)) / (
+        np.power(10, hi) - np.power(10, lo)
+    )
+    return result
+
+
+def scale_power2(array_in, lo=None, hi=None):
+    if lo is None:
+        lo = np.min(array_in)
+    if hi is None:
+        hi = np.max(array_in)
+    validate_for_scaling(array_in, lo, hi)
+    result = 1 / 9 * (np.power(10, (array_in - lo) / (hi - lo)) - 1)
+    return result
+
+
+def unscale_linear(array_in, lo, hi):
+    result = array_in * (hi - lo) / 1.0 + lo
+    return result
+
+
+def unscale_log(array_in, lo, hi):
+    result = lo * np.power(hi / lo, array_in)
+
+    # result = ((np.log10(array_in) - np.log10(lo))
+    #              / (np.log10(hi) - np.log10(lo)))
+    # out = math.pow(lo * (hi / lo), (array_in / 10.0))
+    #                 out = (
+    #                     10
+    #                     * (math.log10(array_in) - math.log10(lo))
+    #                     / (math.log10(hi) - math.log10(lo))
+    #                 )
+    return result
+
+
+def unscale_log2(array_in, lo=None, hi=None):
+    result = (np.power(10, array_in / 1.0) - 1) * (hi - lo) / 9.0 + lo
+    # out = (math.pow(10, array_in / 10.0) - 1) * (
+    #     hi - lo
+    # ) / 9.0 + lo
+
+    return result
+
+
+def unscale_power(array_in, lo, hi):
+    # check if lo and hi were provided
+    # result = np.log10((array_in / 10.0) * (np.power(10, hi) - np.power(10, lo))
+    #                 + np.power(10, lo))
+    result = np.log10(
+        (array_in / 1.0) * (np.power(10, hi) - np.power(10, lo)) + np.power(10, lo)
+    )
+    return result
+
+
+def unscale_power2(array_in, lo, hi):
+    result = np.log10(9.0 * array_in / 1.0 + 1) * (hi - lo) + lo
+    return result
+
+
+class BaseScaler:
```

Comment:
I may be misunderstanding the usage of the annotations here, but what is the outcome of this class? It seems that arrays that are transformed will raise exceptions for any input.

Reply:
I'll add a comment to explain the purpose of the BaseScaler class; transform() and inverse_transform() should be implemented by the derived classes, so it raises an error if called from the base class.